### ISSUE TITLE
Understand the Replayer Drift and Propose Solutions

### ISSUE THREAD

**Issue by paulofelipefeitosa** [vtex/reliability-research#146]
This issue is the result of #97 which gave us the context to know our Replayer problems. The outcome of this issue is to propose and evaluate solutions for the problems found at issue #97.

## Current Behavior
The main problem is the huge difference between the actual and expected requests inter-arrival time.

## Possible Solution
Before the proposal and implementation of any solution, we wish to measure and find the responsible for the huge inter-arrival diff time. 

An initial evaluation methodology is to create a [microbenchmark](https://stackoverflow.com/questions/2842695/what-is-microbenchmarking) that can measure each request inter-arrival time components and that could be used as a performance test for future Replayer modifications.

Also, before implementing any test and evaluation, we would like to have a proof of concept that the evaluation will answer the initial question.

**Comment by paulofelipefeitosa** [comment]
Our replayer quality analysis showed that there is a huge difference between the actual and expected requests inter-arrival time.

Now, I wish to understand where this problem is coming from or at least discard some initial hypotheses. Our initial hypotheses rely on how the Replayer is implemented.

The Replayer implements the Producer and Consumer design pattern:
* The Producer sequentially iterates over all the requests input and puts them into a Channel. However, to try to simulate the request arrival delay, before inserting the request into the Channel the Producer sleeps for the specified inter-arrival time of the request.
* The Consumer’s burden is to recover the requests from the Channel and apply them into a SOLR server waiting for its completion and storing its result. In our experiment, we used a pool of 512 Consumer goroutines.

We have two initial hypotheses:
1. Golang's `time.Sleep` is inaccurate for *microseconds* sleep durations.
2. Golang's channels could be a bottleneck.

To confirm the hypotheses, I intend to write a microbenchmark that will measure the contribution of the `time.Sleep` and the Golang's Channel put and get operations in the actual inter-arrival time for each individual request.

This evaluation could help us to understand the causes of the inter-arrival problem and do not need to be executed for a real workload.

WDYT @thiagomanel @danielfireman?

**Comment by thiagomanel** [comment]
hi @paulofelipefeitosa. 

could you expand on the microbench idea? I think it would be neat to have a micro bench, however, maybe we are talking about different things (I have the gut feeling there is a simpler option to test the time.Sleep hypothesis then constructing a micro bench).

**Comment by paulofelipefeitosa** [comment]
Hi @thiagomanel the idea is to execute a fake workload and calculate the `time.Sleep` actual time and how much time each request spent on the channel.

**Comment by thiagomanel** [comment]
> Hi @thiagomanel the idea is to execute a fake workload and calculate the `time.Sleep` actual time and how much time each request spent on the channel.

do you plan to replace SOLR by some dummy stuff?

**Comment by paulofelipefeitosa** [comment]
@thiagomanel yes, but I am still evaluating if the time to refactor the Replayer to such a thing is worth.

**Comment by thiagomanel** [comment]
Regarding the time.Sleep, I wonder if using a program that executes sleep calls in a loop, and traces the actual and expected delays would be simpler than microbench the replayer (maybe, there is already some public code analyzing this)

But, I see the advantage of having a micro bench, so I am fine with your proposal. When are we going to see the results? :)

**Comment by paulofelipefeitosa** [comment]
For sure on Monday!
